File: .artifacts/deploy-prod-snapshot-20260328b/src/services/emergencyService.js
// --- C:\vibecity.live\src\services\emergencyService.js ---
// Nationwide emergency support based on repo-owned public data.

const EMERGENCY_DATASET_PATH = "/data/emergency-locations.json";
const DEFAULT_LIMIT_PER_TYPE = 3;
const DEFAULT_RADIUS_KM = 80;

let emergencyDatasetCache = null;
let emergencyDatasetPromise = null;

/**
 * Emergency contacts for Thailand
 */
export const EMERGENCY_CONTACTS = {
	police: {
		name: "Police",
		number: "191",
		icon: "Shield",
	},
	touristPolice: {
		name: "Tourist Police",
		number: "1155",
		icon: "BadgeCheck",
	},
	ambulance: {
		name: "Ambulance",
		number: "1669",
		icon: "Ambulance",
	},
	fire: {
		name: "Fire Department",
		number: "199",
		icon: "Flame",
	},
};

const toFiniteNumber = (value) => {
	const parsed = Number(value);
	return Number.isFinite(parsed) ? parsed : null;
};

const normalizeText = (value) => String(value || "").trim();

const buildNearbySearchQuery = (baseQuery, lat, lng) => {
	const safeLat = toFiniteNumber(lat);
	const safeLng = toFiniteNumber(lng);
	if (safeLat === null || safeLng === null) {
		return `${baseQuery} thailand`;
	}
	return `${baseQuery} near ${safeLat.toFixed(6)},${safeLng.toFixed(6)}`;
};

const toRadians = (value) => (Number(value) * Math.PI) / 180;

const calculateDistanceKm = (lat1, lng1, lat2, lng2) => {
	const safeLat1 = toFiniteNumber(lat1);
	const safeLng1 = toFiniteNumber(lng1);
	const safeLat2 = toFiniteNumber(lat2);
	const safeLng2 = toFiniteNumber(lng2);
	if (
		safeLat1 === null ||
		safeLng1 === null ||
		safeLat2 === null ||
		safeLng2 === null
	) {
		return null;
	}

	const earthRadiusKm = 6371;
	const dLat = toRadians(safeLat2 - safeLat1);
	const dLng = toRadians(safeLng2 - safeLng1);
	const a =
		Math.sin(dLat / 2) ** 2 +
		Math.cos(toRadians(safeLat1)) *
			Math.cos(toRadians(safeLat2)) *
			Math.sin(dLng / 2) ** 2;
	const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
	return earthRadiusKm * c;
};

const getEmergencySortOrder = (type) => {
	if (type === "hospital") return 0;
	if (type === "police") return 1;
	if (type === "fire") return 2;
	return 3;
};

const buildEmergencyItem = (item, distance) => ({
	id: item.id,
	name: normalizeText(item.name),
	type: normalizeText(item.type).toLowerCase(),
	lat: toFiniteNumber(item.lat),
	lng: toFiniteNumber(item.lng),
	phone: normalizeText(item.phone) || null,
	province: normalizeText(item.province) || null,
	address: normalizeText(item.address) || null,
	is24h: Boolean(item.is24h),
	distance: Number(distance.toFixed(1)),
	distanceKm: Number(distance.toFixed(2)),
});

export const getEmergencySearchLink = (type, userLat, userLng) => {
	const baseQuery = type === "police" ? "police station" : "hospital emergency";
	return `https://www.google.com/maps/search/?api=1&query=${encodeURIComponent(
		buildNearbySearchQuery(baseQuery, userLat, userLng),
	)}`;
};

export const getEmergencyFallbackState = (userLat, userLng) => ({
	nearest: [],
	hospitals: [],
	police: [],
	searchLinks: {
		hospitals: getEmergencySearchLink("hospital", userLat, userLng),
		police: getEmergencySearchLink("police", userLat, userLng),
	},
});

const loadEmergencyDataset = async () => {
	if (Array.isArray(emergencyDatasetCache)) {
		return emergencyDatasetCache;
	}

	if (!emergencyDatasetPromise) {
		emergencyDatasetPromise = fetch(EMERGENCY_DATASET_PATH, {
			cache: "no-store",
		})
			.then((response) => {
				if (!response.ok) {
					throw new Error(
						`Failed to load emergency dataset (${response.status})`,
					);
				}
				return response.json();
			})
			.then((payload) => {
				const rows = Array.isArray(payload) ? payload : [];
				emergencyDatasetCache = rows;
				return rows;
			})
			.catch(() => {
				emergencyDatasetCache = [];
				return [];
			})
			.finally(() => {
				emergencyDatasetPromise = null;
			});
	}

	return emergencyDatasetPromise;
};

export const getNearbyEmergency = async (
	userLat,
	userLng,
	limitPerType = DEFAULT_LIMIT_PER_TYPE,
	radiusKm = DEFAULT_RADIUS_KM,
) => {
	const safeLat = toFiniteNumber(userLat);
	const safeLng = toFiniteNumber(userLng);
	const fallback = getEmergencyFallbackState(safeLat, safeLng);

	if (safeLat === null || safeLng === null) {
		return fallback;
	}

	const dataset = await loadEmergencyDataset();
	if (!dataset.length) {
		return fallback;
	}

	const safeLimit = Math.max(1, Number(limitPerType) || DEFAULT_LIMIT_PER_TYPE);
	const safeRadius = Math.max(1, Number(radiusKm) || DEFAULT_RADIUS_KM);

	const nearby = dataset
		.map((item) => {
			const distance = calculateDistanceKm(
				safeLat,
				safeLng,
				item?.lat,
				item?.lng,
			);
			if (distance === null || distance > safeRadius) return null;
			return buildEmergencyItem(item, distance);
		})
		.filter(Boolean)
		.sort((left, right) => {
			if (left.distance !== right.distance) {
				return left.distance - right.distance;
			}
			return (
				getEmergencySortOrder(left.type) - getEmergencySortOrder(right.type)
			);
		});

	if (!nearby.length) {
		return fallback;
	}

	const hospitals = nearby
		.filter((item) => item.type === "hospital")
		.slice(0, safeLimit);
	const police = nearby
		.filter((item) => item.type === "police")
		.slice(0, safeLimit);

	return {
		nearest: nearby.slice(0, safeLimit * 2),
		hospitals,
		police,
		searchLinks: fallback.searchLinks,
	};
};

/**
 * Generate tel: link for calling
 */
export const getCallLink = (number) => {
	return `tel:${String(number || "").replaceAll(/\s+/g, "")}`;
};

/**
 * Generate navigation link to emergency location
 */
export const getDirectionsLink = (lat, lng) => {
	return `https://www.google.com/maps/dir/?api=1&destination=${lat},${lng}&travelmode=driving`;
};

File: .artifacts/deploy-prod-snapshot-20260328b/backend/app/services/emergency_service.py
"""
Emergency Service - Find nearby hospitals, police stations, and fire stations
from the seeded Thailand-wide emergency_locations dataset.
"""

import asyncio
import math
from dataclasses import dataclass

from app.core.supabase import supabase, supabase_admin


@dataclass
class EmergencyLocation:
    id: str
    name: str
    lat: float
    lng: float
    type: str  # hospital, police, fire_station
    phone: str | None = None
    distance_km: float | None = None
    opening_hours: str | None = None
    emergency: bool = True


EMERGENCY_CONTACTS = {
    "police": {"name": "ตำรวจ", "number": "191", "icon": "🚔"},
    "tourist_police": {"name": "ตำรวจท่องเที่ยว", "number": "1155", "icon": "👮"},
    "ambulance": {"name": "รถพยาบาล", "number": "1669", "icon": "🚑"},
    "fire": {"name": "ดับเพลิง", "number": "199", "icon": "🚒"},
    "highway_police": {"name": "ตำรวจทางหลวง", "number": "1193", "icon": "🛣️"},
    "rescue": {"name": "กู้ชีพ/กู้ภัย", "number": "1554", "icon": "⛑️"},
}


def calculate_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Haversine formula for distance in km."""
    radius = 6371
    dlat = math.radians(lat2 - lat1)
    dlng = math.radians(lng2 - lng1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlng / 2) ** 2
    )
    c = 2 * math.asin(math.sqrt(a))
    return radius * c


class EmergencyService:
    """Service to find nearby emergency facilities from seeded DB data."""

    def __init__(self):
        self._client = supabase_admin or supabase

    def _fetch_seed_rows(self, location_type: str | None = None) -> list[dict]:
        if not self._client:
            return []

        try:
            query = self._client.table("emergency_locations").select(
                "id,name,type,latitude,longitude,phone,address,province,is_24h"
            )
            if location_type:
                query = query.eq("type", location_type)
            response = query.execute()
            return getattr(response, "data", None) or []
        except Exception:
            return []

    async def _get_seed_rows(self, location_type: str | None = None) -> list[dict]:
        return await asyncio.to_thread(self._fetch_seed_rows, location_type)

    def _parse_seed_row(
        self,
        row: dict,
        user_lat: float,
        user_lng: float,
    ) -> EmergencyLocation | None:
        try:
            lat = float(row.get("latitude"))
            lng = float(row.get("longitude"))
        except (TypeError, ValueError):
            return None

        distance = calculate_distance(user_lat, user_lng, lat, lng)
        location_type = str(row.get("type") or "").strip().lower() or "unknown"

        return EmergencyLocation(
            id=str(row.get("id", "")),
            name=str(row.get("name") or f"Unknown {location_type}").strip(),
            lat=lat,
            lng=lng,
            type=location_type,
            phone=str(row.get("phone") or "").strip() or None,
            distance_km=round(distance, 2),
            opening_hours="24/7" if row.get("is_24h") else None,
            emergency=True,
        )

    async def _get_nearby_by_type(
        self,
        lat: float,
        lng: float,
        location_type: str,
        radius_m: int,
        limit: int,
    ) -> list[EmergencyLocation]:
        rows = await self._get_seed_rows(location_type)
        radius_km = max(radius_m / 1000, 1)
        locations = []

        for row in rows:
            parsed = self._parse_seed_row(row, lat, lng)
            if not parsed or parsed.distance_km is None:
                continue
            if parsed.distance_km <= radius_km:
                locations.append(parsed)

        locations.sort(key=lambda item: item.distance_km or 999)
        return locations[:limit]

    async def get_nearby_hospitals(
        self,
        lat: float,
        lng: float,
        radius_m: int = 10000,
        limit: int = 5,
    ) -> list[EmergencyLocation]:
        return await self._get_nearby_by_type(lat, lng, "hospital", radius_m, limit)

    async def get_nearby_police(
        self,
        lat: float,
        lng: float,
        radius_m: int = 10000,
        limit: int = 5,
    ) -> list[EmergencyLocation]:
        return await self._get_nearby_by_type(lat, lng, "police", radius_m, limit)

    async def get_nearby_fire_stations(
        self,
        lat: float,
        lng: float,
        radius_m: int = 15000,
        limit: int = 3,
    ) -> list[EmergencyLocation]:
        return await self._get_nearby_by_type(lat, lng, "fire", radius_m, limit)

    async def get_all_nearby(
        self,
        lat: float,
        lng: float,
        radius_m: int = 10000,
    ) -> dict[str, list[EmergencyLocation]]:
        hospitals, police, fire = await asyncio.gather(
            self.get_nearby_hospitals(lat, lng, radius_m),
            self.get_nearby_police(lat, lng, radius_m),
            self.get_nearby_fire_stations(lat, lng, radius_m),
        )

        return {
            "hospitals": hospitals,
            "police": police,
            "fire_stations": fire,
            "emergency_contacts": list(EMERGENCY_CONTACTS.values()),
        }

    def get_emergency_contacts(self) -> list[dict]:
        return [{**value, "type": key} for key, value in EMERGENCY_CONTACTS.items()]


emergency_service = EmergencyService()
